cranelift/docs/isle-integration.md

```diff
@@ -0,0 +1,188 @@
+aarch64, we implement the handful of move mitosis special cases at the
+`inst.isle` layer. Either way, the important thing is that the lowering rules
+remain pure.
+
+Fnally, note that these moves are generally cleaned up by the register
+allocator's move coalescing, and move mitosis will eventually go away completely
+once we switch over to `regalloc2`, which takes instructions in SSA form
+directly as input.
+
+Instructions that implicitly operate on specific registers, or which require
+that certain operands be in certain registers, are handled similarly: the
+lowering rules use a pure paradigm that ignores these constraints and has
+instructions that explicitly take implicit operands, and we ensure the
+constraints are fulfilled a layer below the lowering rules (in `inst.isle` or in
+Rust glue code).
+
+## When are lowering rules allowed to have side effects?
+
+Extractors (the matchers that appear on the left-hand sides of `rule`s) should
+**never** have side effects. When evaluating a rule's extractors, we haven't yet
+committed to evaluating that rule's right-hand side. If the extractors performed
+side effects, we could get deeply confusing action-at-a-distance bugs where
+rules we never fully match pull the rug out from under our feet.
+
+Anytime you are tempted to perform side effects in an extractor, you should
+instead just package up the things you would need in order to perform that side
+effect, and then have a separate constructor that takes that package and
+performs the side effect it describes. The constructor can only be called inside
+a rule's right-hand side, which is only evaluated after we've committed to this
+rule, which avoids the action-at-a-distance bugs described earlier.
+
+For example, loads have a side effect in CLIF: they might trap. Therefore, even
+if a loaded value is never used, we will emit code that implements that
+load. But if we are compiling for x86 we can sink loads into other the operand
+for another operation depending on how the loaded value is used. If we sink that
+load into, say, an `add` then we need to tell the lowering context *not* to
+lower the CLIF `load` instruction anymore, because its effectively already
+lowered as part of lowering the `add` that uses the loaded value. That is a side
+effect, and we might be tempted to perform that side effect in the extractor
+that matches sinkable loads. But we can't do that because although the load
+itself might be sinkable, there might be a reason why we ultimately don't
+perform this load-sinking rule, and if that happens we still need to lower the
+CLIF load. So we make the `sinkable_load` extractor create a `SinkableLoad` type
+that packages up everything we need to know about the load and how to tell the
+lowering context that we've sunk it and the lowering context doesn't need to
+lower it anymore, but *it doesn't actually tell that to the lowering context
+yet*. Then, we pair that with a `sink_load` constructor that takes the
```

Comment:
The description of what constitutes a side effect, I am having a hard time appreciating what's really being described or taking it an applying it and thinking up a different example. Is there some way or mechanism to help prevent writing code that violates this no side effect rule? With the description of a side effect here it worries me that it will be easier to create bugs that are harder to track down.

Reply:
I rewrote this paragraph a bit and broke it up with examples, as Chris suggested. Hope it is clearer now.

>  It has me thinking, is there some way or mechanism to help prevent writing code that violates this no side effect rule? With the description of a side effect here it worries me that it will be easier to create bugs that are harder to track down.

It's the same thing you have to deal with writing Rust code right now too. If you do

```rust
if something_with_side_effects() {
    foo();
} else {
    bar();
}
```

the side effects of the condition aren't rolled back if you take the `bar();` branch.

Unfortunately, this isn't really something we can reason about in ISLE itself because all the things that could perform side effects are external Rust helpers.